src/p11_child/p11_child_openssl.c

```diff
@@ -0,0 +1,1133 @@
+                        &signature_size);
+    if (rv != CKR_OK) {
+        DEBUG(SSSDBG_OP_FAILURE, "C_Sign failed.\n");
+        return EIO;
+    }
+
+    cert_pub_key = X509_get_pubkey(cert->cert);
+    if (cert_pub_key == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "X509_get_pubkey failed.\n");
+        ret = EIO;
+        goto done;
+    }
+
+    md_ctx = EVP_MD_CTX_create();
+    if (md_ctx == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "EVP_MD_CTX_create failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+    ret = EVP_VerifyInit(md_ctx, EVP_sha1());
+    if (ret != 1) {
+        DEBUG(SSSDBG_OP_FAILURE, "EVP_VerifyInit failed.\n");
+        ret = EINVAL;
+        goto done;
+    }
+
+    ret = EVP_VerifyUpdate(md_ctx, random_value, sizeof(random_value));
+    if (ret != 1) {
+        DEBUG(SSSDBG_OP_FAILURE, "EVP_VerifyUpdate failed.\n");
+        ret = EINVAL;
+        goto done;
+    }
+
+    ret = EVP_VerifyFinal(md_ctx, signature, signature_size, cert_pub_key);
+    if (ret != 1) {
+        DEBUG(SSSDBG_OP_FAILURE, "EVP_VerifyFinal failed.\n");
+        ret = EINVAL;
+        goto done;
+    }
+
+    ret = EOK;
+
+done:
+    talloc_free(signature);
+    EVP_PKEY_free(cert_pub_key);
+
+    return ret;
+}
+
+#define MAX_SLOTS 64
+
+errno_t do_card(TALLOC_CTX *mem_ctx, struct p11_ctx *p11_ctx,
+                enum op_mode mode, const char *pin,
+                const char *module_name_in, const char *token_name_in,
+                const char *key_id_in, char **_multi)
+{
+    int ret;
+    size_t c;
+    size_t s;
+    CK_FUNCTION_LIST **modules;
+    CK_FUNCTION_LIST *module = NULL;
+    char *mod_name;
+    char *mod_file_name;
+    CK_ULONG num_slots;
+    CK_SLOT_ID slots[MAX_SLOTS];
+    CK_SLOT_ID slot_id;
+    CK_SLOT_INFO info;
+    CK_TOKEN_INFO token_info;
+    CK_RV rv;
+    size_t module_id;
+    char *module_file_name = NULL;
+    char *slot_name = NULL;
+    char *token_name = NULL;
+    CK_SESSION_HANDLE session = 0;
+    struct cert_list *cert_list = NULL;
+    struct cert_list *item = NULL;
+    char *multi = NULL;
+    bool pkcs11_session = false;
+    bool pkcs11_login = false;
+
+    /* Maybe use P11_KIT_MODULE_TRUSTED ? */
+    modules = p11_kit_modules_load_and_initialize(0);
+    if (modules == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "p11_kit_modules_load_and_initialize failed.\n");
+        return EIO;
+    }
+
+    DEBUG(SSSDBG_TRACE_ALL, "Module List:\n");
+    for (c = 0; modules[c] != NULL; c++) {
+        mod_name = p11_kit_module_get_name(modules[c]);
+        mod_file_name = p11_kit_module_get_filename(modules[c]);
+        DEBUG(SSSDBG_TRACE_ALL, "common name: [%s].\n", mod_name);
+        DEBUG(SSSDBG_TRACE_ALL, "dll name: [%s].\n", mod_file_name);
+        free(mod_name);
+        free(mod_file_name);
+
+        num_slots = MAX_SLOTS;
+        rv = modules[c]->C_GetSlotList(CK_TRUE, slots, &num_slots);
```

Comment:
What is a good documentation to understand what these functions do? I read https://p11-glue.github.io/p11-glue/p11-kit/manual/index.html but that only tells me what the p11_kit_* functions do..

Reply:
Those are the PKCS#11 calls, see e.g. http://docs.oasis-open.org/pkcs11/pkcs11-base/v2.40/pkcs11-base-v2.40.pdf for details.